src/fulfillment/src/main/java/mushop/FulfillmentService.java

```diff
@@ -0,0 +1,101 @@
+import io.nats.client.Dispatcher;
+import io.nats.client.Message;
+import io.nats.client.Nats;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Singleton;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+@Singleton
+@Introspected
+public class FulfillmentService {
+    private Connection nc;
+    private final ObjectMapper objectMapper = new ObjectMapper();
+    private static final Logger LOG = LoggerFactory.getLogger(FulfillmentService.class);
+    @Value("nats://${mushop.messaging.host}:${mushop.messaging.port}")
+    private String NATS_URL;
+    @Value("${mushop.messaging.subjects.orders}")
+    private String MUSHOP_ORDERS_SUBJECT;
+    @Value("${mushop.messaging.subjects.shipments}")
+    private String MUSHOP_SHIPMENTS_SUBJECT;
+    @Value("${mushop.messaging.simulation-delay}")
+    private Long simulationDelay;
+
+    public FulfillmentService() {
+    }
+
+    @EventListener
+    @Async
+    public void connect(final ServiceStartedEvent event) throws InterruptedException {
+        LOG.info("Connecting to NATS {} and subscribing to subject {}", this.NATS_URL, this.MUSHOP_ORDERS_SUBJECT);
+        boolean connected = false;
+        while (!connected) {
+            try {
+                Future<Boolean> result = Executors.newSingleThreadExecutor().submit(() -> {
+                    try {
+                        nc = Nats.connect(this.NATS_URL);
+                        Dispatcher d = nc.createDispatcher((msg) -> {
+                            try {
+                                handleMessage(msg);
+                            } catch (IOException e) {
+                                e.printStackTrace();
+                            }
+                        });
+                        d.subscribe(this.MUSHOP_ORDERS_SUBJECT);
+                        return Boolean.TRUE;
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                        LOG.error("Connection failed due to {}. Retrying in 5s", e.getMessage());
+                        Thread.sleep(5000l);
+                        return Boolean.FALSE;
+                    }
+                });
+                connected = result.get();
+            } catch (ExecutionException e) {
+                e.printStackTrace();
+            }
+        }
+        LOG.info("Connected to NATS {} and subscribed to subject {}", this.NATS_URL, this.MUSHOP_ORDERS_SUBJECT);
+
+    }
+
+    private void handleMessage(Message message) throws IOException {
+        String response = new String(message.getData(), StandardCharsets.UTF_8);
+        OrderUpdate update = objectMapper.readValue(message.getData(), OrderUpdate.class);
+        LOG.info("got message {} on the mushop orders subject", update);
+        fulfillOrder(update);
```

Comment:
might be cleaner is `handleMessage` returned `OrderUpdate` and then you separately call `fulfillOrder` on line 55. (but might be a Java thing as well :))

Reply:
Yes, this may make more sense to read. 
I did it this way mostly to structure the exception handling the way I have it. I will try this out and see.

Reply:
Yes this is better and more readable and the exception handling does not change. refactored.